src/mol-math/geometry/symmetry-operator.ts

```diff
@@ -167,127 +168,123 @@ namespace SymmetryOperator {
-        const invariantPosition = createCoordinateMapper(SymmetryOperator.Default, coords);
-        const position = operator.isIdentity ? invariantPosition : createCoordinateMapper(operator, coords);
-        const { x, y, z } = createProjections(operator, coords);
-        return { operator, coordinates: coords, invariantPosition, position, x, y, z, r: radius };
-    }
+        invariantPosition(i: T, s: Vec3): Vec3 {
+            s[0] = this._x[i];
+            s[1] = this._y[i];
+            s[2] = this._z[i];
+            return s;
+        }
 
-    export function createMapping<T extends number>(operator: SymmetryOperator, coords: Coordinates, radius: ((index: T) => number) = _zeroRadius) {
-        return _createMapping(operator, coords, radius);
-    }
+        position(i: T, s: Vec3): Vec3 {
+            s[0] = this._x[i];
+            s[1] = this._y[i];
+            s[2] = this._z[i];
+            Vec3.transformMat4(s, s, this._m);
+            return s;
+        }
 
-    export function createCoordinateMapper<T extends number>(t: SymmetryOperator, coords: Coordinates): CoordinateMapper<T> {
-        if (t.isIdentity) return identityPosition(coords);
-        return generalPosition(t, coords);
-    }
-}
+        x(i: T): number {
+            const m = this._m;
+            const xx = m[0], yy = m[4], zz = m[8], tx = m[12];
 
-export { SymmetryOperator };
+            const x = this._x[i], y = this._y[i], z = this._z[i], w = (m[3] * x + m[7] * y + m[11] * z + m[15]) || 1.0;
+            return (xx * x + yy * y + zz * z + tx) / w;
+        }
 
-function _zeroRadius(i: number) { return 0; }
+        y(i: T): number {
+            const m = this._m;
+            const xx = m[1], yy = m[5], zz = m[9], ty = m[13];
 
-interface Projections { x(index: number): number, y(index: number): number, z(index: number): number }
+            const x = this._x[i], y = this._y[i], z = this._z[i], w = (m[3] * x + m[7] * y + m[11] * z + m[15]) || 1.0;
+            return (xx * x + yy * y + zz * z + ty) / w;
+        }
 
-function createProjections(t: SymmetryOperator, coords: SymmetryOperator.Coordinates): Projections {
-    if (t.isIdentity) return { x: projectCoord(coords.x), y: projectCoord(coords.y), z: projectCoord(coords.z) };
-    return { x: projectX(t, coords), y: projectY(t, coords), z: projectZ(t, coords) };
-}
+        z(i: T): number {
+            const m = this._m;
+            const xx = m[2], yy = m[6], zz = m[10], tz = m[14];
 
-function projectCoord(xs: ArrayLike<number>) {
-    return function projectCoord(i: number) {
-        return xs[i];
-    };
-}
+            const x = this._x[i], y = this._y[i], z = this._z[i], w = (m[3] * x + m[7] * y + m[11] * z + m[15]) || 1.0;
+            return (xx * x + yy * y + zz * z + tz) / w;
+        }
+    }
 
-function isW1(m: Mat4) {
-    return m[3] === 0 && m[7] === 0 && m[11] === 0 && m[15] === 1;
-}
+    class _ArrayMappingW1<T extends number> implements ArrayMapping<T> {
```

Comment:
Add `_ArrayMappingIdentity` as well that skips the matrix transforms altogether?

Reply:
good idea, done